sql/core/src/main/scala/org/apache/spark/sql/DataFrameReader.scala

```diff
@@ -402,7 +402,11 @@ class DataFrameReader private[sql](sparkSession: SparkSession) extends Logging {
    * @since 2.0.0
    */
   @scala.annotation.varargs
-  def json(paths: String*): DataFrame = format("json").load(paths : _*)
+  def json(paths: String*): DataFrame = {
+    userSpecifiedSchema.foreach(
+      ExprUtils.checkJsonSchema(_).foreach(e => throw new AnalysisException(e)))
```

Comment:
I think we would have to throw an exception via `QueryCompilationErrors`, cc @karenfeng

Reply:
Ideally, yes; and there would be an error class associated with this rather than an arbitrary error message. This one may be a little harder because we're using `TypeCheckFailure` with the same message. @allisonwang-db, what do you think?

Reply:
I see, thenI am fine without fixing it here but let me leave it to you guys.

Reply:
One possible solution, until we have a more generic solution for `TypeCheckFailure`, is modify checkJsonSchema to return the exception instead of the message `def checkJsonSchema(schema: DataType): Option[AnalysisException]` and for `TypeCheckFailure` get the message from the exception, and in other cases throw the exception.
I'm looking how to add a new exception to `QueryCompilationErrors` properly to push the code with the possible solution.